discoverd2/server/state.go

```diff
@@ -0,0 +1,314 @@
+package server
+
+import (
+	"container/list"
+	"errors"
+	"sync"
+
+	"github.com/flynn/flynn/pkg/stream"
+)
+
+type EventKind uint
+
+const (
+	EventKindUp EventKind = 1 << iota
+	EventKindUpdate
+	EventKindDown
+	EventKindLeader
+)
+
+func (k EventKind) String() string {
+	switch k {
+	case EventKindUp:
+		return "up"
+	case EventKindUpdate:
+		return "update"
+	case EventKindDown:
+		return "down"
+	case EventKindLeader:
+		return "leader"
+	default:
+		return "unknown"
+	}
+}
+
+type Event struct {
+	Service string
+	Kind    EventKind
+	*Instance
+}
+
+func eventKindUpdate(existing bool) EventKind {
+	if existing {
+		return EventKindUpdate
+	}
+	return EventKindUp
+}
+
+type Instance struct {
+	ID     string            `json:"id"`
+	Addr   string            `json:"addr"`
+	Proto  string            `json:"proto"`
+	Meta   map[string]string `json:"meta,omitempty"`
+	TTL    int               `json:"ttl"`
+	Leader bool              `json:"leader,omitempty"`
+	Index  uint              `json:"index"`
+}
+
+func (inst *Instance) Equal(other *Instance) bool {
+	return inst.Addr == other.Addr &&
+		inst.Proto == other.Proto &&
+		inst.Index == other.Index &&
+		mapEqual(inst.Meta, other.Meta)
+}
+
+func mapEqual(x, y map[string]string) bool {
```

Comment:
I think I'll leave it, there's no way the performance of reflect can match this implementation, and it gets called a lot.